### PR TITLE
Abstracts and generics support fixes

### PR DESCRIPTION
**Pull request by mayakwd**
- Add anonymous types in anonymous types support
- Add anonymous structures names support
- Added chained anonymous fields recognition
- Add proper generics propagation for anonymous structures, typedefs and classes
- Add generic constraint support
- Add Null&lt;T&gt; support
- Add proper from-to declaration support, types infer and compatibility checks
- Fix referenced var-init support check

This is main fixes for this PR.

### CODE REVIEW

File: testData/goto/TypeDef2.hx [hunk 1]
```diff
@@ -1,7 +1,7 @@
-typedef ExtendedString<T> = {>String, key:T};
+typedef ExtendedString<T> = {key:T};
```
Comment (EricBishton) [inline testData/goto/TypeDef2.hx]: Why did you change this test?  It used to test typedef extensions (e.g. all fields of String were available in ExtendedString).

Hmm.  I guess it didn't because it was using 'key' instead of some value from String.

I just want to be sure that we still have tests that use the "inherited field" syntax.

File: src/common/com/intellij/plugins/haxe/lang/psi/impl/HaxePsiFieldImpl.java [hunk 2]
```diff
@@ -158,6 +158,17 @@ public PsiType getType() {
     return psiType != null ? psiType : HaxePsiTypeAdapter.DYNAMIC;
   }
 
+  @Override
+  @Nullable
+  public HaxeTypeTag getTypeTag() {
+    return null;
```
[the remark this answers is not in the document]
Comment (mayakwd) [reply src/common/com/intellij/plugins/haxe/lang/psi/impl/HaxePsiFieldImpl.java]: This is the default implementation, everything implemented in child classes. It was done to make PsiField generic, because most of child classes have such methods (excluding enum values fro example)

File: testData/goto/TypeDef2.hx (hunk 1 above)
Comment (mayakwd) [reply testData/goto/TypeDef2.hx]: I've changed it because String is not anonymous type, and it can't be used as parent type.
I'll check tests.

File: src/common/com/intellij/plugins/haxe/model/type/SpecificHaxeClassReference.java [hunk 3]
```diff
@@ -50,34 +61,39 @@ public SpecificHaxeClassReference(
     this.rangeConstraint = rangeConstraint;
   }
 
-  public HaxeClassReference getHaxeClassRef() {
-    return this.clazz;
+  public static SpecificHaxeClassReference withoutGenerics(@NotNull HaxeClassReference clazz) {
+    return new SpecificHaxeClassReference(clazz, ResultHolder.EMPTY, null, null, clazz.elementContext);
```
[the remark this answers is not in the document]
Comment (mayakwd) [reply src/common/com/intellij/plugins/haxe/model/type/SpecificHaxeClassReference.java]: I think it should be made as builder instead of current implementation, but not in this PR. 
It could be tricky a bit.

File: src/common/com/intellij/plugins/haxe/model/type/SpecificHaxeClassReference.java [hunk 4]
```diff
@@ -182,4 +183,167 @@ public ResultHolder access(String name, HaxeExpressionEvaluatorContext context)
...
+    if (stack.contains(model.haxeClass)) return list;
+    stack.push(model.haxeClass);
+
+    if (!model.isAbstract()) {
+      if (model.haxeClass instanceof HaxeTypedefDeclaration) {
+        SpecificHaxeClassReference type = ((AbstractHaxeTypeDefImpl)model.haxeClass).getTargetClass(genericResolver);
+        if (type != null) {
+          list.add(type);
+          list.addAll(type.getCompatibleTypes());
+        }
```
[the remark this answers is not in the document]
Comment (mayakwd) [reply src/common/com/intellij/plugins/haxe/model/type/SpecificHaxeClassReference.java]: Well, they are not exactly the same. I'm not sure if it could be encapsulated in way of better looking. Right now I can't say it's possible to make it better.

File: src/common/com/intellij/plugins/haxe/lang/psi/impl/HaxeReferenceImpl.java [hunk 5]
```diff
@@ -103,6 +105,16 @@ public int getTextOffset() {
   @Nullable
   public HaxeGenericSpecialization getSpecialization() {
     if (LOG.isTraceEnabled()) LOG.trace(traceMsg(null));
+    HaxeGenericSpecialization specialization = getUserData(SPECIALIZATION_KEY);
+    if (specialization == null) {
+      specialization = getSpecializationInternal();
+      putUserData(SPECIALIZATION_KEY, specialization);
+    }
```
[the remark this answers is not in the document]
Comment (mayakwd) [reply src/common/com/intellij/plugins/haxe/lang/psi/impl/HaxeReferenceImpl.java]: Well I'll try to find a possibility to re-validate specialization in future, right now i'll discard this change

File: testData/goto/TypeDef2.hx (hunk 1 above)
Comment (mayakwd) [reply testData/goto/TypeDef2.hx]: I've reverted test, but it's not valid for real, anonymous type can't extend extern classes.

**Comment by mayakwd** [comment]
Well i've done with fixing, review it please
